matrix-bot/src/Network/Matrix/Bot/Async.hs

```diff
@@ -0,0 +1,89 @@
+module Network.Matrix.Bot.Async ( AsyncHandler
+                                , asyncHandler
+                                , SyncGroup
+                                , SyncGroupCall
+                                , syncCall
+                                , asyncGroupCall
+                                , syncGroupCall
+                                , startSyncGroup
+                                ) where
+
+import Control.Concurrent.Async ( Async
+                                , async
+                                , cancel
+                                , wait
+                                )
+import Control.Concurrent.STM (atomically)
+import Control.Concurrent.STM.TBMQueue ( TBMQueue
+                                       , newTBMQueueIO
+                                       , readTBMQueue
+                                       , writeTBMQueue
+                                       )
+import Control.Concurrent.MVar ( MVar
+                               , putMVar
+                               )
+import Control.Monad.IO.Class ( MonadIO
+                              , liftIO
+                              )
+import Control.Monad.IO.Unlift ( UnliftIO
+                               , unliftIO
+                               )
+import Control.Monad.Trans.Resource ( MonadResource
+                                    , ReleaseKey
+                                    , allocate
+                                    )
+import Data.Function (fix)
+
+import Network.Matrix.Bot.State
+
+data SyncGroup (m :: * -> *) a = SyncGroup (Async ()) (TBMQueue (a, Maybe (MVar ()))) ReleaseKey
+
+data SyncGroupCall m where
+  AsyncGroupCall :: SyncGroup m a -> a -> Maybe (MVar ()) -> SyncGroupCall m
+  SyncCall :: (a -> m ()) -> a -> SyncGroupCall m
+
+newtype AsyncHandler m a = AsyncHandler (TBMQueue (a, Maybe (MVar ())) -> m ())
+
+asyncHandler :: (MonadIO n)
+             => (forall b. n b -> m b)
+             -> (a -> n ())
+             -> AsyncHandler m a
+asyncHandler runT handler = AsyncHandler $ \q -> runT $ fix $ \recurse ->
+  liftIO (atomically (readTBMQueue q))
+  >>= maybe (pure ()) (\(e, done) -> handler e *> signalDone done *> recurse)
+  where signalDone Nothing = pure ()
+        signalDone (Just done) = liftIO $ putMVar done ()
+
+syncCall :: (a -> m ()) -> a -> SyncGroupCall m
+syncCall = SyncCall
+
+asyncGroupCall :: SyncGroup m a -> a -> Maybe (MVar ()) -> SyncGroupCall m
+asyncGroupCall = AsyncGroupCall
+
+syncGroupCall :: (MatrixBotBase m) => SyncGroupCall m -> m ()
+syncGroupCall (SyncCall f x) = f x
+syncGroupCall (AsyncGroupCall (SyncGroup _ chan _) x done) = liftIO $ atomically $ writeTBMQueue chan (x, done)
+
+startSyncGroup :: (MonadIO n, MonadResource m)
+               => UnliftIO n
+               -> AsyncHandler n a -> m (SyncGroup n a)
+startSyncGroup unliftHandler (AsyncHandler handler) = do
+  queue <- liftIO $ newTBMQueueIO 20
+  (releaseKey, thread) <- resourceTAsync unliftHandler $ handler queue
```

Comment:
I actually intended to use SyncGroups for this, because the concept is already general enough to cover this use cases. One of the recent commits added `routeAsyncEvent` which can be used to handle a single event asynchronously, but completely hides the fact that sync groups are involved.

That commit also fixed the memory leak I was concerned about. Basically we keep a set of all started sync groups and regularly check if any of any of them stopped and we need to free them.